### ISSUE TITLE
Appending to dataclass list uses last item as defaults

### ISSUE THREAD

**Issue by cifkao** [omni-us/jsonargparse#552]
## 🐛 Bug report

When the type of an argument is list of dataclass instances, the first item in the list uses the dataclass's defaults as expected, but appending further items always uses the last item's values as defaults instead of using the actual defaults.

### To reproduce

`test.py`:
```python
from dataclasses import dataclass
import jsonargparse

@dataclass
class A:
    x: int = 1
    foo: str = "bar"

def main(a: list[A] = None) -> None:
    print(a)

jsonargparse.CLI(main)
```

Command:
```
python test.py --a+='{}' --a+='{"x":2, "foo": "spam"}' --a+='{"x": 3}'
```

Output (`foo='spam'` is copied from the 2nd item to the 3rd):
```python
[A(x=1, foo='bar'), A(x=2, foo='spam'), A(x=3, foo='spam')]
```

### Expected behavior

Expected output (3rd item uses the default `foo='bar'`):
```python
[A(x=1, foo='bar'), A(x=2, foo='spam'), A(x=3, foo='bar')]
```

### Environment

- jsonargparse version: 4.31.0
- Python version: 3.10.13
- How jsonargparse was installed: `pip install jsonargparse`
- OS: Linux

**Comment by mauvilsa** [comment]
Thank you for reporting!

**Comment by mauvilsa** [comment]
This should be fixed with #560. Feel free to try it out.
